lib/internal/crypto/random.js

```diff
@@ -164,36 +171,59 @@ function randomInt(min, max, callback) {
   const excess = RAND_MAX % range;
   const randLimit = RAND_MAX - excess;
 
-  if (isSync) {
-    // Sync API
-    while (true) {
-      const x = randomBytes(6).readUIntBE(0, 6);
-      // If x > (maxVal - (maxVal % range)), we will get "modulo bias"
-      if (x > randLimit) {
-        // Try again
-        continue;
-      }
+  // If we don't have a callback, or if there is still data in the cache, we can
+  // do this synchronously, which is super fast.
+  while (isSync || (randomCacheOffset < randomCache.length)) {
+    if (randomCacheOffset === randomCache.length) {
+      // This might block the thread for a bit, but we are in sync mode.
+      randomFillSync(randomCache);
+      randomCacheOffset = 0;
+    }
+
+    const x = randomCache.readUIntBE(randomCacheOffset, 6);
+    randomCacheOffset += 6;
+
+    // If x <= (maxVal - (maxVal % range)), we have avoided "modulo bias".
+    if (x <= randLimit) {
       const n = (x % range) + min;
-      return n;
+      if (isSync) return n;
+      setImmediate(() => callback(undefined, n));
```

Comment:
While correct, I would recommend using `process.nextTick` in this case. Deferring for a full event loop cycle when the data is already available is usually a source of performance loss.